src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +199,231 @@ bool intersects(Ray const &ray, Box const &box)
   return intersection(ray, box, tmin, tmax) && (tmax >= 0.f);
 }
 
+// The function returns the index of the largest
+// component of the direction vector.
+KOKKOS_INLINE_FUNCTION int findLargestComp(Vector const &dir)
+{
+  int kz = 0;
+  for (int i = 1; i < 3; i++)
+  {
+    float compmax = std::fabs(dir[i - 1]);
+    if (std::fabs(dir[i]) > compmax)
+    {
+      compmax = dir[i];
+      kz = i;
+    }
+  }
+  return kz;
+}
+
+// Both the ray and the triangle were transformed beforehand
+// so that the ray is a unit vector along the z-axis (0,0,1),
+// and the triangle is transformed with the same matrix
+// (which is M in the paper). This function is called only
+// when the ray is co-planar to the triangle (with the
+// determinant being zero). The rotation by this function is
+// to prepare for the ray-edge intersection calculations in 2D.
+// The rotation is around the z-axis. For any point after
+// the rotation, its new x* equals its original length
+// with the correct sign, and the new y* = z. The current
+// implementation avoids explicitly defining rotation angles
+// and directions. The following ray-edge intersection will
+// be in the x*-y* plane.
+KOKKOS_INLINE_FUNCTION Point rotate2D(Point const &point)
+{
+  Point point_star;
+  float r = std::sqrt(point[0] * point[0] + point[1] * point[1]);
+  if (point[0] != 0)
+  {
+    point_star[0] = (point[0] > 0 ? 1 : -1) * r;
+  }
+  else
+  {
+    point_star[0] = (point[1] > 0 ? 1 : -1) * r;
+  }
+  point_star[1] = point[2];
+  point_star[2] = 0.0;
+  return point_star;
+}
+
+// The function is for ray-edge intersection
+// with the rotated ray along the z-axis and
+// the transformed and rotated triangle edges
+// The algorithm is described in
+// https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection#Given_two_points_on_each_line_segment
+KOKKOS_INLINE_FUNCTION bool rayEdgeIntersect(Point const &edge_vertex_1,
+                                             Point const &edge_vertex_2,
+                                             float &t)
+{
+  float x3 = edge_vertex_1[0];
+  float y3 = edge_vertex_1[1];
+  float x4 = edge_vertex_2[0];
+  float y4 = edge_vertex_2[1];
+
+  float y2;
+  if (KokkosExt::min(y3, y4) >= 0.f)
+  {
+    y2 = KokkosExt::max(y3, y4);
+  }
+  else
+  {
+    y2 = KokkosExt::min(y3, y4);
+  }
+
+  float det = y2 * (x3 - x4);
+
+  auto const epsilon = 0.00001f;
```

Comment:
Magic constant. Would be nice to explain why we need it, and what should be a proper value.

Reply:
It depends on the size of the problem as briefly discussed in the previous PR.
See the new tests and the new epsilon added in the commit.

Reply:
If I further increase the size range, the current epsilon (for u v w and det) will fail.

Reply:
Bruno found in the previous PR that it was due to precisions.
I think the roll-back precision in this algorithm doesn't work as it claimed to be, at least the roll-back itself needs an epsilon to check if it is needed.
But if you check the numbers when scaling the problem in size, the algorithm is not scaling with it with the precisions/if checks.